File: sdks/python/legacybaas/services/blockchain_router.py
"""
Blockchain Router Service
Smart routing across 11+ blockchain networks
"""

from typing import Dict, Any, List, Optional
from ..errors import LegacyBaaSError, BlockchainError


class BlockchainRouter:
    """Blockchain Router for Python SDK"""
    
    def __init__(self, client):
        self.client = client
    
    def route_transaction(
        self,
        transaction: Dict[str, Any],
        preferences: Optional[Dict[str, Any]] = None
    ) -> Dict[str, Any]:
        """
        Route transaction using AI-powered smart routing
        
        Args:
            transaction: Blockchain transaction data
            preferences: Routing preferences
            
        Returns:
            Transaction result
        """
        default_preferences = {
            "preferred_networks": [],
            "max_fee": None,
            "max_time": None,
            "prioritize_speed": False,
            "prioritize_cost": False,
            "prioritize_security": True
        }
        
        if preferences:
            default_preferences.update(preferences)
        
        try:
            return self.client._make_request(
                "POST",
                "/blockchain/route",
                data={
                    "transaction": transaction,
                    "preferences": default_preferences
                }
            )
        except Exception as e:
            raise BlockchainError(f"Transaction routing failed: {e}")
    
    def get_routing_recommendation(
        self,
        transaction: Dict[str, Any],
        preferences: Optional[Dict[str, Any]] = None
    ) -> Dict[str, Any]:
        """
        Get optimal routing recommendation
        
        Args:
            transaction: Transaction to analyze
            preferences: Routing preferences
            
        Returns:
            Routing recommendations with AI insights
        """
        return self.client._make_request(
            "POST",
            "/blockchain/route/recommend",
            data={
                "transaction": transaction,
                "preferences": preferences or {}
            }
        )
    
    def get_transaction_status(self, transaction_id: str) -> Dict[str, Any]:
        """
        Get transaction status
        
        Args:
            transaction_id: Transaction ID
            
        Returns:
            Transaction status
        """
        return self.client._make_request(
            "GET",
            f"/blockchain/transactions/{transaction_id}"
        )
    
    def track_transaction(self, blockchain_tx_id: str) -> Dict[str, Any]:
        """
        Track transaction by blockchain hash
        
        Args:
            blockchain_tx_id: Blockchain transaction ID
            
        Returns:
            Transaction tracking info
        """
        return self.client._make_request(
            "GET",
            f"/blockchain/track/{blockchain_tx_id}"
        )
    
    def get_supported_networks(self) -> Dict[str, Any]:
        """
        Get supported networks
        
        Returns:
            Supported blockchain networks
        """
        return self.client._make_request(
            "GET",
            "/blockchain/networks"
        )
    
    def get_network_info(self, network_id: str) -> Dict[str, Any]:
        """
        Get specific network information
        
        Args:
            network_id: Network identifier
            
        Returns:
            Network information
        """
        return self.client._make_request(
            "GET",
            f"/blockchain/networks/{network_id}"
        )
    
    def get_network_health(self) -> Dict[str, Any]:
        """
        Get network health metrics
        
        Returns:
            Network health status
        """
        return self.client._make_request(
            "GET",
            "/blockchain/health"
        )
    
    def estimate_gas(
        self,
        transaction: Dict[str, Any],
        network_id: Optional[str] = None
    ) -> List[Dict[str, Any]]:
        """
        Estimate gas fees for transaction
        
        Args:
            transaction: Transaction data
            network_id: Specific network to estimate for
            
        Returns:
            Gas estimates
        """
        return self.client._make_request(
            "POST",
            "/blockchain/estimate-gas",
            data={
                "transaction": transaction,
                "network_id": network_id
            }
        )
    
    def create_wallet(self, wallet_data: Dict[str, Any]) -> Dict[str, Any]:
        """
        Create new wallet
        
        Args:
            wallet_data: Wallet creation data
            
        Returns:
            Wallet information
        """
        return self.client._make_request(
            "POST",
            "/blockchain/wallets",
            data=wallet_data
        )
    
    def get_wallet_balance(self, address: str, network: str) -> Dict[str, Any]:
        """
        Get wallet balance
        
        Args:
            address: Wallet address
            network: Network identifier
            
        Returns:
            Wallet balance information
        """
        return self.client._make_request(
            "GET",
            f"/blockchain/balance/{address}",
            headers={"X-Network": network}
        )
    
    def list_wallets(self, filters: Optional[Dict[str, Any]] = None) -> Dict[str, Any]:
        """
        List wallets
        
        Args:
            filters: Optional filters
            
        Returns:
            Wallet list with pagination
        """
        return self.client._make_request(
            "GET",
            "/blockchain/wallets",
            params=filters or {}
        )
    
    def get_cross_chain_bridge_status(self) -> Dict[str, Any]:
        """
        Get cross-chain bridge status
        
        Returns:
            Bridge status information
        """
        return self.client._make_request(
            "GET",
            "/blockchain/bridges"
        )
    
    def get_metrics(self) -> Dict[str, Any]:
        """
        Get blockchain metrics
        
        Returns:
            Blockchain operation metrics
        """
        return self.client._make_request(
            "GET",
            "/blockchain/metrics"
        )

File: sdks/typescript/src/services/BlockchainRouter.ts
/**
 * Blockchain Router Service
 * Smart routing across 11+ blockchain networks
 */

import { LegacyBaaSClient } from '../client/LegacyBaaSClient';
import { ApiResponse } from '../types';
import { LegacyBaaSError } from '../errors';

export interface BlockchainTransaction {
  transactionId: string;
  amount: string;
  currency: string;
  fromAddress: string;
  toAddress: string;
  network: string;
  metadata?: Record<string, any>;
  priority?: 'low' | 'normal' | 'high' | 'urgent';
  swiftReference?: string;
}

export interface RoutingPreferences {
  preferredNetworks?: string[];
  maxFee?: number;
  maxTime?: number;
  prioritizeSpeed?: boolean;
  prioritizeCost?: boolean;
  prioritizeSecurity?: boolean;
}

export interface TransactionResult {
  transactionId: string;
  blockchainTxId: string;
  network: string;
  status: 'pending' | 'confirmed' | 'failed' | 'routing';
  fees: {
    networkFee: number;
    platformFee: number;
    totalFee: number;
    currency: string;
  };
  timestamps: {
    submitted: string;
    confirmed?: string;
    finalized?: string;
  };
  confirmations: number;
  blockNumber?: number;
  estimatedTime?: string;
}

export interface NetworkInfo {
  networkId: string;
  name: string;
  type: 'mainnet' | 'testnet' | 'layer2';
  status: 'active' | 'maintenance' | 'deprecated';
  capabilities: {
    smartContracts: boolean;
    multiSig: boolean;
    crossChain: boolean;
    privacyFeatures: boolean;
  };
  fees: {
    avgTransactionFee: number;
    avgConfirmationTime: number;
    currency: string;
  };
  metrics: {
    tps: number;
    uptime: number;
    totalTransactions: number;
  };
}

export interface WalletData {
  address: string;
  network: string;
  balance: string;
  currency: string;
  isActive: boolean;
  createdAt: string;
  metadata?: Record<string, any>;
}

export interface GasEstimate {
  network: string;
  estimatedGas: number;
  gasPrice: number;
  totalCost: number;
  currency: string;
  confidence: number;
  estimatedTime: string;
}

export class BlockchainRouter {
  constructor(private client: LegacyBaaSClient) {}

  /**
   * Route transaction using AI-powered smart routing
   */
  async routeTransaction(
    transaction: BlockchainTransaction,
    preferences: RoutingPreferences = {}
  ): Promise<TransactionResult> {
    const response = await this.client.makeRequest<TransactionResult>(
      'POST',
      '/blockchain/route',
      {
        transaction,
        preferences: {
          preferredNetworks: preferences.preferredNetworks || [],
          maxFee: preferences.maxFee,
          maxTime: preferences.maxTime,
          prioritizeSpeed: preferences.prioritizeSpeed || false,
          prioritizeCost: preferences.prioritizeCost || false,
          prioritizeSecurity: preferences.prioritizeSecurity || true
        }
      }
    );

    return response.data;
  }

  /**
   * Get optimal routing recommendation
   */
  async getRoutingRecommendation(
    transaction: BlockchainTransaction,
    preferences: RoutingPreferences = {}
  ): Promise<{
    recommendations: Array<{
      network: string;
      score: number;
      estimatedFee: number;
      estimatedTime: string;
      pros: string[];
      cons: string[];
    }>;
    aiInsights: {
      marketConditions: string;
      networkCongestion: Record<string, number>;
      recommendation: string;
      confidence: number;
    };
  }> {
    const response = await this.client.makeRequest<{
      recommendations: Array<{
        network: string;
        score: number;
        estimatedFee: number;
        estimatedTime: string;
        pros: string[];
        cons: string[];
      }>;
      aiInsights: {
        marketConditions: string;
        networkCongestion: Record<string, number>;
        recommendation: string;
        confidence: number;
      };
    }>('POST', '/blockchain/route/recommend', {
      transaction,
      preferences
    });

    return response.data;
  }

  /**
   * Get transaction status
   */
  async getTransactionStatus(transactionId: string): Promise<TransactionResult> {
    const response = await this.client.makeRequest<TransactionResult>(
      'GET',
      `/blockchain/transactions/${transactionId}`
    );

    return response.data;
  }

  /**
   * Track transaction by blockchain hash
   */
  async trackTransaction(blockchainTxId: string): Promise<{
    transactionId: string;
    blockchainTxId: string;
    network: string;
    status: string;
    confirmations: number;
    blockNumber?: number;
    gasUsed?: number;
    gasFee?: number;
    timestamp: string;
    details: Record<string, any>;
  }> {
    const response = await this.client.makeRequest<{
      transactionId: string;
      blockchainTxId: string;
      network: string;
      status: string;
      confirmations: number;
      blockNumber?: number;
      gasUsed?: number;
      gasFee?: number;
      timestamp: string;
      details: Record<string, any>;
    }>('GET', `/blockchain/track/${blockchainTxId}`);

    return response.data;
  }

  /**
   * Get supported networks
   */
  async getSupportedNetworks(): Promise<{
    networks: NetworkInfo[];
    totalNetworks: number;
    activeNetworks: number;
  }> {
    const response = await this.client.makeRequest<{
      networks: NetworkInfo[];
      totalNetworks: number;
      activeNetworks: number;
    }>('GET', '/blockchain/networks');

    return response.data;
  }

  /**
   * Get specific network information
   */
  async getNetworkInfo(networkId: string): Promise<NetworkInfo> {
    const response = await this.client.makeRequest<NetworkInfo>(
      'GET',
      `/blockchain/networks/${networkId}`
    );

    return response.data;
  }

  /**
   * Get network health metrics
   */
  async getNetworkHealth(): Promise<{
    networks: Record<string, {
      status: string;
      uptime: number;
      avgResponseTime: number;
      errorRate: number;
      lastChecked: string;
    }>;
    overallHealth: number;
  }> {
    const response = await this.client.makeRequest<{
      networks: Record<string, {
        status: string;
        uptime: number;
        avgResponseTime: number;
        errorRate: number;
        lastChecked: string;
      }>;
      overallHealth: number;
    }>('GET', '/blockchain/health');

    return response.data;
  }

  /**
   * Estimate gas fees for transaction
   */
  async estimateGas(
    transaction: BlockchainTransaction,
    networkId?: string
  ): Promise<GasEstimate[]> {
    const response = await this.client.makeRequest<GasEstimate[]>(
      'POST',
      '/blockchain/estimate-gas',
      {
        transaction,
        networkId
      }
    );

    return response.data;
  }

  /**
   * Create new wallet
   */
  async createWallet(walletData: {
    network: string;
    walletType: 'personal' | 'business' | 'escrow';
    metadata?: Record<string, any>;
  }): Promise<{
    address: string;
    network: string;
    walletType: string;
    publicKey: string;
    createdAt: string;
    walletId: string;
  }> {
    const response = await this.client.makeRequest<{
      address: string;
      network: string;
      walletType: string;
      publicKey: string;
      createdAt: string;
      walletId: string;
    }>('POST', '/blockchain/wallets', walletData);

    return response.data;
  }

  /**
   * Get wallet balance
   */
  async getWalletBalance(
    address: string,
    network: string
  ): Promise<{
    address: string;
    network: string;
    balances: Array<{
      asset: string;
      balance: string;
      usdValue: number;
      lastUpdated: string;
    }>;
    totalUsdValue: number;
  }> {
    const response = await this.client.makeRequest<{
      address: string;
      network: string;
      balances: Array<{
        asset: string;
        balance: string;
        usdValue: number;
        lastUpdated: string;
      }>;
      totalUsdValue: number;
    }>('GET', `/blockchain/balance/${address}`, undefined, {
      headers: { 'X-Network': network }
    });

    return response.data;
  }

  /**
   * List wallets
   */
  async listWallets(filters: {
    network?: string;
    walletType?: string;
    isActive?: boolean;
    limit?: number;
    offset?: number;
  } = {}): Promise<{
    wallets: WalletData[];
    pagination: {
      total: number;
      limit: number;
      offset: number;
      hasMore: boolean;
    };
  }> {
    const response = await this.client.makeRequest<{
      wallets: WalletData[];
      pagination: {
        total: number;
        limit: number;
        offset: number;
        hasMore: boolean;
      };
    }>('GET', '/blockchain/wallets', undefined, {
      headers: { 'X-Query-Params': JSON.stringify(filters) }
    });

    return response.data;
  }

  /**
   * Get cross-chain bridge status
   */
  async getCrossChainBridgeStatus(): Promise<{
    bridges: Array<{
      bridgeId: string;
      fromNetwork: string;
      toNetwork: string;
      status: 'active' | 'maintenance' | 'disabled';
      avgFee: number;
      avgTime: string;
      supportedAssets: string[];
    }>;
    totalBridges: number;
    activeBridges: number;
  }> {
    const response = await this.client.makeRequest<{
      bridges: Array<{
        bridgeId: string;
        fromNetwork: string;
        toNetwork: string;
        status: 'active' | 'maintenance' | 'disabled';
        avgFee: number;
        avgTime: string;
        supportedAssets: string[];
      }>;
      totalBridges: number;
      activeBridges: number;
    }>('GET', '/blockchain/bridges');

    return response.data;
  }

  /**
   * Get blockchain metrics
   */
  async getMetrics(): Promise<{
    totalTransactions: number;
    successRate: number;
    avgProcessingTime: number;
    networkDistribution: Record<string, number>;
    volumeByNetwork: Record<string, number>;
    gasSavings: number;
  }> {
    const response = await this.client.makeRequest<{
      totalTransactions: number;
      successRate: number;
      avgProcessingTime: number;
      networkDistribution: Record<string, number>;
      volumeByNetwork: Record<string, number>;
      gasSavings: number;
    }>('GET', '/blockchain/metrics');

    return response.data;
  }
}